0.5.3 erroring on CI with ValueError: No threading layer could be loaded.

*Sorry for the long report, but I've been adding to it while debugging!*

Just started getting test failures on CI, which seem to be coming from pynndescent 0.5.3. Downgrading pynndescent stops the errors. It seems to be related to the threading backend:

```pytb
E           numba.core.errors.LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E           No threading layer could be loaded.
E           HINT:
E           Intel TBB is required, try:
E           $ conda/pip install tbb
E           
E           File "../../../../../g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/umap/umap_.py", line 404:
E           def compute_membership_strengths(
E               <source elided>
E           
E               rows = np.zeros(knn_indices.size, dtype=np.int32)
E               ^
E           
E           During: lowering "id=20[LoopNest(index_variable = parfor_index.936, range = (0, $28load_attr.11, 1))]{344: <ir.Block at /g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/umap/umap_.py (404)>}Var(parfor_index.936, umap_.py:404)" at /g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/umap/umap_.py (404)
```

[More logs from CI](https://gist.github.com/ivirshup/48321b8c2282748f874597f7f649d475)

I would note that I do not normally install `tbb` on these builds, but numba seems to detect it's availability:

<details>
<summary> Threading info from `numba -s` on our CI </summary>

```
__Threading Layer Information__
TBB Threading Layer Available                 : True
+-->TBB imported successfully.
OpenMP Threading Layer Available              : True
+-->Vendor: GNU
Workqueue Threading Layer Available           : True
+-->Workqueue imported successfully.
```

But `tbb` isn't installed

```
__Installed Packages__
_libgcc_mutex             0.1                        main  
brotlipy                  0.7.0           py38h27cfd23_1003  
ca-certificates           2020.10.14                    0  
certifi                   2020.6.20          pyhd3eb1b0_3  
cffi                      1.14.3           py38h261ae71_2  
chardet                   3.0.4           py38h06a4308_1003  
conda                     4.9.2            py38h06a4308_0  
conda-package-handling    1.7.2            py38h03888b9_0  
cryptography              3.2.1            py38h3c74f83_1  
idna                      2.10                       py_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20191231         h14c3975_1  
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.1.0                hdf63c60_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
ncurses                   6.2                  he6710b0_1  
openssl                   1.1.1h               h7b6447c_0  
pip                       20.2.4           py38h06a4308_0  
pycosat                   0.6.3            py38h7b6447c_1  
pycparser                 2.20                       py_2  
pyopenssl                 19.1.0             pyhd3eb1b0_1  
pysocks                   1.7.1            py38h06a4308_0  
python                    3.8.5                h7579374_1  
readline                  8.0                  h7b6447c_0  
requests                  2.24.0                     py_0  
ruamel_yaml               0.15.87          py38h7b6447c_1  
setuptools                50.3.1           py38h06a4308_1  
six                       1.15.0           py38h06a4308_0  
sqlite                    3.33.0               h62c20be_0  
tk                        8.6.10               hbc83047_0  
tqdm                      4.51.0             pyhd3eb1b0_0  
urllib3                   1.25.11                    py_0  
wheel                     0.35.1             pyhd3eb1b0_0  
xz                        5.2.5                h7b6447c_0  
yaml                      0.2.5                h7b6447c_0  
zlib                      1.2.11               h7b6447c_3  
```

If I do explicitly install `tbb`

</details>

# Sorta able to reproduce locally

### Can't reproduce on macOS

I'm not able to reproduce this on my MacBook, but it looks like numba has some handling around osx from the tracebacks. I'd note I haven't been able to install tab on my mac for a while. 

### Able to reproduce on linux (but there may be more environment issues preventing a fix)

I'm able to reproduce the errors on a local HPC, but installing `tbb` does not solve the issue. There may be a conflict with the environment though, since the hpc has tbb and I get errors about the `tbb` protocol being out of date:

```python
In [1]: import pynndescent
   ...: import numpy as np
   ...: 
   ...: X = np.random.random_sample((10_000, 50))
   ...: 
   ...: nn = pynndescent.NNDescent(X)
/g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/numba/np/ufunc/parallel.py:365: NumbaWarning: The TBB threading layer requires TBB version 2019.5 or later i.e., TBB_INTERFACE_VERSION >= 11005. Found TBB_INTERFACE_VERSION = 10002. The TBB threading layer is disabled.
  warnings.warn(problem)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-61fda4417201> in <module>
      4 X = np.random.random_sample((10_000, 50))
      5 
----> 6 nn = pynndescent.NNDescent(X)

/g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/pynndescent/pynndescent_.py in __init__(self, data, metric, metric_kwds, n_neighbors, n_trees, leaf_size, pruning_degree_multiplier, diversify_prob, n_search_trees, tree_init, init_graph, random_state, low_memory, max_candidates, n_iters, delta, n_jobs, compressed, verbose)
    775 
    776         # Set threading constraints
--> 777         self._original_num_threads = numba.get_num_threads()
    778         if self.n_jobs != -1 and self.n_jobs is not None:
    779             numba.set_num_threads(self.n_jobs)

/g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/numba/np/ufunc/parallel.py in get_num_threads()
    616 
    617     """
--> 618     _launch_threads()
    619     num_threads = _get_num_threads()
    620     if num_threads <= 0:

/g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/numba/np/ufunc/parallel.py in _launch_threads()
    497             # No lib found, raise and hint
    498             if not lib:
--> 499                 raise_with_hint(requirements)
    500 
    501             ll.add_symbol('numba_parallel_for', lib.parallel_for)

/g/data3/yf0/user/iv5408/miniconda3/envs/scanpy-dev2/lib/python3.8/site-packages/numba/np/ufunc/parallel.py in raise_with_hint(required)
    442                     options = '\nOR\n'.join([err_helpers[x] for x in required])
    443                     hint = hintmsg % ("One of:\n%s" % options)
--> 444                 raise ValueError(errmsg % hint)
    445 
    446             if t in namedbackends:

ValueError: No threading layer could be loaded.
HINT:
Intel TBB is required, try:
$ conda/pip install tbb
```

But `pip list | grep tbb` says: 

```
tbb               2021.3.0
```

In this environment, if I change the threading_layer to not be `tbb`, the previous example works. E.g.:

<details>
<summary> </summary>

```python
import numba
import pynndescent
import numpy as np

# Either other threading backend works
# numba.config.THREADING_LAYER = "workqueue"
numba.config.THREADING_LAYER = "omp"

X = np.random.random_sample((10_000, 50))

nn = pynndescent.NNDescent(X)
```

</details>

### Can't reproduce on a old linux server where tbb can't be installed

On an older centos 7 server I'm not able to reproduce an y errors, but `numba -s` reports that `tbb` can't be found there.

# Summary

I'm unsure of whether I'm running into one or multiple bugs. One commonality is that I cannot seem to get a working version of the tbb library when I try, but in some cases this gets by the import check:

https://github.com/lmcinnes/pynndescent/blob/c16d0594584cb3d4ce1f8ddd6698f21debfba74b/pynndescent/__init__.py#L6-L12

Ideally, I would like to have full control over setting the backend, and not have the possibility of importing a package changing that. However it does seem like there is an upstream issue around numba being able to accurately tell whether a backend is available.

* Is there some reason `omp` doesn't work here? 
* Why the switch to preferring `tbb`? Would it be possible to revert this change?

--- a/pynndescent/__init__.py
+++ b/pynndescent/__init__.py
@@ -3,12 +3,13 @@ import numba
 from .pynndescent_ import NNDescent, PyNNDescentTransformer
 
 # Workaround: https://github.com/numba/numba/issues/3341
-try:
-    from numba.np.ufunc import tbbpool
+if numba.config.THREADING_LAYER == 'omp':
+    try:
+        from numba.np.ufunc import tbbpool
 
-    numba.config.THREADING_LAYER = "tbb"
-except ImportError as e:
-    # might be a missing symbol due to e.g. tbb libraries missing
-    numba.config.THREADING_LAYER = "workqueue"
+        numba.config.THREADING_LAYER = "tbb"
+    except ImportError as e:
+        # might be a missing symbol due to e.g. tbb libraries missing
+        numba.config.THREADING_LAYER = "workqueue"
 
 __version__ = pkg_resources.get_distribution("pynndescent").version

--- a/pynndescent/__init__.py
+++ b/pynndescent/__init__.py
@@ -3,7 +3,7 @@ import numba
 from .pynndescent_ import NNDescent, PyNNDescentTransformer
 
 # Workaround: https://github.com/numba/numba/issues/3341
-if numba.config.THREADING_LAYER == 'omp':
+if numba.config.THREADING_LAYER == "omp":
     try:
         from numba.np.ufunc import tbbpool
 

